src/middleware.js

```diff
@@ -2,11 +2,37 @@ import path from 'path';
 
 import mime from 'mime-types';
 
+import getETag from 'etag';
+
 import getFilenameFromUrl from './utils/getFilenameFromUrl';
 import handleRangeHeaders from './utils/handleRangeHeaders';
 import ready from './utils/ready';
 
+const { hasOwnProperty } = Object.prototype;
+
 export default function wrapper(context) {
+  const { compiler } = context;
+  const etagRegistry = new Map();
```

Comment:
memory leak

Reply:
We need to `WeakMap` and use `assets` source as key

Reply:
AFAIK, Weakmaps don't allow primitives (eg. strings) to be used as keys

Reply:
No, it is still memory leak, you should not use primitives for weak amp

Reply:
Resolved the other thread about this so all conversation is in this thread.

Can you explain why there would still be a memory leak if unused etags are being cleaned up per build?

Reply:
> Can you explain why there would still be a memory leak if unused etags are being cleaned up per build?

Because we create `etagRegistry` on middleware start so next compilation will put new files again and again, and if you use `name.[contenthash][ext]` you will increase size infinitely, it will be difficult to debug, but this is a potential leak, using `WeakMap` help to solve it, when asset was removed from `comilation.assets` it will be removed from `WeakMap`, so we free memory

Reply:
I see, thanks for explaining.

That's addressed by [`etagRegistry.clear()`](https://github.com/webpack/webpack-dev-middleware/pull/746/files#diff-9a117d52a8a2cb3e5f498575bdfdea4dc32418180c09029454fbaf7f95ebe5a1R19) on every build completion because `.clear()` removes all entries in a Map.


I'm wondering how you imagine using a WeakMap is possible? I would jump to using it if it were possible to get the asset object (they key) when a new HTTP request comes in [here](https://github.com/webpack/webpack-dev-middleware/pull/746/files#diff-9a117d52a8a2cb3e5f498575bdfdea4dc32418180c09029454fbaf7f95ebe5a1R77).

Reply:
> That's addressed by etagRegistry.clear() on every build completion because .clear() removes all entries in a Map.

it was reduce performance, we should avoid it

> I'm wondering how you imagine using a WeakMap is possible? I would jump to using it if it were possible to get the asset object (they key) when a new HTTP request comes in here.

store source of assets and use it as key